pkg/operator/backupcontroller/backupcontroller.go

```diff
@@ -0,0 +1,364 @@
+	backup.Status.Conditions = append(backup.Status.Conditions, v1.Condition{
+		Type:               "BackupCompleted",
+		Reason:             string(operatorv1alpha1.BackupSkipped),
+		Message:            "skipped due too many simultaneous backups",
+		Status:             v1.ConditionTrue,
+		LastTransitionTime: metav1.NewTime(time.Now()),
+	})
+
+	_, err := client.UpdateStatus(ctx, &backup, v1.UpdateOptions{})
+	return err
+}
+
+func indexJobsByBackupLabelName(jobs *batchv1.JobList) map[string]batchv1.Job {
+	m := map[string]batchv1.Job{}
+	if jobs == nil {
+		return m
+	}
+
+	for _, j := range jobs.Items {
+		backupCrdName := j.Labels[backupJobLabel]
+		if backupCrdName != "" {
+			m[backupCrdName] = *j.DeepCopy()
+		}
+	}
+
+	return m
+}
+
+func findRunningJobs(jobs *batchv1.JobList) []batchv1.Job {
+	var running []batchv1.Job
+	if jobs == nil {
+		return running
+	}
+
+	for _, j := range jobs.Items {
+		if !isJobFinished(&j) {
+			running = append(running, *j.DeepCopy())
+		}
+	}
+
+	return running
+}
+
+// isJobFinished checks whether the given Job has finished execution.
+// It does not discriminate between successful and failed terminations.
+func isJobFinished(j *batchv1.Job) bool {
+	for _, c := range j.Status.Conditions {
+		if (c.Type == batchv1.JobComplete || c.Type == batchv1.JobFailed) && c.Status == corev1.ConditionTrue {
+			return true
+		}
+	}
+	return false
+}
+
+func reconcileJobStatus(ctx context.Context,
+	jobClient batchv1client.JobInterface,
+	backupClient operatorv1alpha1client.EtcdBackupInterface,
+	job batchv1.Job,
+	backup operatorv1alpha1.EtcdBackup) error {
+
+	jobState := batchv1.JobConditionType("")
+	for _, c := range job.Status.Conditions {
+		// the types and type transitions are compatible between jobs and our backup states
+		if (c.Type == batchv1.JobComplete || c.Type == batchv1.JobFailed) && c.Status == corev1.ConditionTrue {
+			jobState = c.Type
+			break
+		}
+	}
+
+	// we only reconcile completed/failed jobs
+	if string(jobState) == "" {
+		return nil
+	}
+
+	if backup.Labels == nil {
+		backup.Labels = map[string]string{}
+	}
+
+	backup.Labels["state"] = "processed"
+
+	// we're updating the backup and its status first, in case of a failure we can still reconcile the job without it
+	_, err := backupClient.Update(ctx, &backup, v1.UpdateOptions{})
+	if err != nil {
+		return fmt.Errorf("error while updating backup [%s]: %w", backup.Name, err)
+	}
+
+	// update above has changed the CRD, to avoid conflicts we grab the latest version from the API
+	bp, err := backupClient.Get(ctx, backup.Name, v1.GetOptions{})
+	if err != nil {
+		return fmt.Errorf("error while getting backup for updating status [%s]: %w", backup.Name, err)
+	}
+
+	statusReason := operatorv1alpha1.BackupCompleted
+	if jobState == batchv1.JobFailed {
+		statusReason = operatorv1alpha1.BackupFailed
+	}
+
+	bp.Status.Conditions = append(bp.Status.Conditions,
+		v1.Condition{
+			Type:               "BackupCompleted",
```

Comment:
Ah okay I get it. I initially only defined the condition types.
But we're only using `BackupCompleted` as the Type with the reasons `BackupFailed` `BackupSkipped`, and (a little confusingly) `BackupCompleted`.

Now that I see this in practice I should have done a better job of maybe defining the successful condition with Type as `BackupCompleted` with the reason as `BackupSucceeded`.

I can update the condition types later but WDYT about making `BackupSucceeded` or something similar as the reason for a successful completion?

Reply:
oh shoot, I got confused because the type was called "...Reason":
https://github.com/openshift/api/pull/1482/files#diff-07061d3f7cd529f689aeef37c7e2cc30938d36ee18a91fdbb8f456585bb582a8R82

I personally would stay with the definition of the batchv1/job:

> if (c.Type == batchv1.JobComplete || c.Type == batchv1.JobFailed)

BackupCompleted means success, BackupFailed is failure. 
Do we need to switch the reason with the type or do you prefer to have the reason be the same as the type?